crates/telio-traversal/src/endpoint_state.rs

```diff
@@ -0,0 +1,120 @@
+use std::fmt;
+use telio_utils::telio_log_warn;
+
+#[derive(Debug, Default, PartialEq)]
+pub enum EndpointState {
+    #[default]
+    Disconnected,
+    EndpointGathering,
+    Ping,
+    Published,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Event {
+    SendCallMeMaybeRequest,
+    ReceiveCallMeMaybeResponse,
+    Publish,
+    Timeout,
+    EndpointGone,
+}
+
+#[derive(Default, Debug)]
+pub struct EndpointStateMachine {
+    state: EndpointState,
+    last_event: Option<Event>,
+}
+
+impl EndpointStateMachine {
+    pub fn handle_event(&mut self, event: Event) {
+        match (&self.state, &event) {
+            (EndpointState::Disconnected, Event::SendCallMeMaybeRequest) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::EndpointGathering;
+            }
+
+            (EndpointState::EndpointGathering, Event::ReceiveCallMeMaybeResponse) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Ping;
+            }
+
+            (EndpointState::Ping, Event::Publish) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Published;
+            }
+
+            (EndpointState::EndpointGathering, Event::Timeout)
+            | (EndpointState::Ping, Event::Timeout) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Disconnected;
+            }
+
+            (EndpointState::Published, Event::EndpointGone) => {
+                self.last_event = Some(event);
+                self.state = EndpointState::Disconnected;
+            }
+
+            (_, event) => {
+                telio_log_warn!("Invalid state transition {:?} -> {:?}", &self.state, event);
+            }
+        }
+    }
+
+    pub fn get(&self) -> &EndpointState {
+        &self.state
+    }
+
+    pub fn last_event(&self) -> &Option<Event> {
+        &self.last_event
+    }
```

Comment:
Because then these two don't have to have the reference (at least I'm 95% sure that `Option<T>` is `Copy` when `T` is `Copy`

Reply:
agreed